src/devices/machine/w5100.cpp

```diff
@@ -0,0 +1,4320 @@
+					m_interrupt_timer->adjust(clocks_to_attotime(intptmr * 4));
+		}
+
+		m_irq_handler(m_irq_state ? ASSERT_LINE : CLEAR_LINE);
+	}
+
+}
+
+int w5100_base_device::register_bank(uint16_t offset)
+{
+	if ((offset & 0x0700) < 0x0400) return 0;
+
+	if (m_device_type == dev_type::W5100)
+	{
+		if (offset > 0x07ff) return 7;
+	}
+	return (offset & 0x0700) >> 8;
+}
+
+
+/*
+ * Direct bus interface: 15-bit address, 8-bit data bus
+ * Indirect bus interface: 2-bit address, 8-bit data bus
+ * W5100s is always indirect w/ auto-increment
+ */
+
+uint8_t w5100_base_device::read(uint16_t offset)
+{
+	if (m_common_registers[0] & MR_IND)
+	{
+		switch(offset & 0x03)
+		{
+			case IDM_OR:
+				return m_common_registers[0];
+				break;
+			case IDM_AR0:
+				return m_idm >> 8;
+				break;
+			case IDM_AR1:
+				return m_idm & 0xff;
+				break;
+			case IDM_DR:
+				offset = m_idm;
+
+				if ((m_common_registers[0] & MR_AI) && !machine().side_effects_disabled())
+				{
+					m_idm++;
+					if (m_device_type == dev_type::W5100 && (m_idm & 0x1fff) == 0x0000)
+						m_idm -= 0x2000;
+				}
+				break;
+		}
+	}
+
+	offset &= 0x7fff;
+
+	return m_address_space->read_byte(offset);
+}
+
+
+void w5100_base_device::write(uint16_t offset, uint8_t data)
+{
+	if (m_common_registers[0] & MR_IND)
+	{
+		switch(offset & 0x03)
+		{
+			case IDM_OR:
+				offset = 0;
+				break;
+			case IDM_AR0:
+				m_idm = (m_idm & 0x00ff) | (data << 8);
+				return;
+				break;
+			case IDM_AR1:
+				m_idm = (m_idm & 0xff00) | data;
+				return;
+				break;
+			case IDM_DR:
+				offset = m_idm;
+				if (m_common_registers[0] & MR_AI)
+				{
+					m_idm++;
+					if (m_device_type == dev_type::W5100 && (m_idm & 0x1fff) == 0x0000)
+						m_idm -= 0x2000;
+				}
+				break;
+		}
+	}
+
+
+	offset &= 0x7fff;
+	LOGMASKED(LOG_WRITE, "write(0x%04x, 0x%02x)\n", offset, data);
+
+	m_address_space->write_byte(offset, data);
+}
+
+uint8_t w5100_base_device::read_socket_register_3(offs_t offset) const
+{
+	return read_socket_register(0x0300 | (offset & 0xff));
+}
+
+
+uint8_t w5100_base_device::read_socket_register(offs_t offset) const
+{
+	int sn = offset >> 8;
+	offset &= 0xff;
+	if (offset >= 0x80) return 0x00;
+
+	return m_socket_registers[sn][offset];
+}
+
+void w5100_base_device::write_socket_register_3(offs_t offset, uint8_t data)
+{
+	write_socket_register(0x0300 | (offset & 0xff), data);
+}
+
+void w5100_base_device::write_socket_register(offs_t offset, uint8_t data)
+{
+	const bool W5100S = m_device_type == dev_type::W5100S;
+
+	int sn = offset >> 8;
+	offset &= 0xff;
+
+	uint8_t *socket = m_socket_registers[sn];
+
+	switch(offset)
+	{
+		case Sn_MR:
+			socket[offset] = data;
+			if ((sn == 0) && (data & Sn_MR_MF) == Sn_MR_MF)
+				set_promisc(false);
+			break;
+
+		case Sn_PORT0:
+		case Sn_PORT1:
+		case Sn_PROTO:
+		case Sn_TOS:
+		case Sn_TTL:
+		case Sn_FRAGR0:
+		case Sn_FRAGR1:
+			/* these update immediately */
+			socket[offset] = data;
+			break;
```

Comment:
If you’re using big switch statements inside your address space handlers, it usually means you have a problem.  You should be getting the memory system to do the heavy lifting for you and using on handler for each logical function.